README
MooseX-Traits-Attribute-CascadeClear

A simple set of attribute traits that can be used to cascade "clear"
operations.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc MooseX::Traits::Attribute::CascadeClear

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-Traits-Attribute-CascadeClear

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/MooseX-Traits-Attribute-CascadeClear

    CPAN Ratings
        http://cpanratings.perl.org/d/MooseX-Traits-Attribute-CascadeClear

    Search CPAN
        http://search.cpan.org/dist/MooseX-Traits-Attribute-CascadeClear/


COPYRIGHT AND LICENSE

Copyright (C) 2008 Chris Weyl <cweyl@alumni.drew.edu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the 
    
    Free Software Foundation, Inc. 
    59 Temple Place, Suite 330
    Boston, MA  02111-1307  USA



Changes
Revision history for MooseX-Traits-Attribute-CascadeClear

0.03    Sat Jan 17 12:13:04 PST 2009
- add a minimum version of Moose to Makefile.PL (0.64); this should address
  some tester failures.

0.02    Mon Dec 29 20:39:31 PST 2008
- correct MooseX::Traits::Attribute::CascadeClearMaster pod

0.01    Sun Dec 28 16:24:47 PST 2008
- initial release



t/lib/DoubleCascade.pm
#############################################################################
#
# Test using CascadeClear traits... 
#
# Author:  Chris Weyl (cpan:RSRCHBOY), <cweyl@alumni.drew.edu>
# Company: No company, personal work
# Created: 12/28/2008 09:42:30 AM PST
#
# Copyright (c) 2008 Chris Weyl <cweyl@alumni.drew.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
#############################################################################

package DoubleCascade;

use Moose;

extends 'SingleCascade';

has sub4 => (
    traits => [ 'CascadeClear', 'CascadeClearMaster' ],

    clear_on => 'master',

    is => 'ro',
    isa => 'Str',
    lazy_build => 1,
);

sub _build_sub4 { shift->master . '4' }

has sub4_sub1 => (
    traits => [ 'CascadeClear' ],

    clear_on => 'sub4',

    is => 'ro',
    isa => 'Str',
    lazy_build => 1,
);

sub _build_sub4_sub1 { shift->sub4 . 'sub1' }

1;

__END__

=head1 NAME

<Module::Name> - <One line description of module's purpose>

=head1 VERSION

The initial template usually just has:

This documentation refers to <Module::Name> version 0.0.1


=head1 SYNOPSIS

	use <Module::Name>;
	# Brief but working code example(s) here showing the most common usage(s)

	# This section will be as far as many users bother reading
	# so make it as educational and exemplary as possible.


=head1 DESCRIPTION

A full description of the module and its features.
May include numerous subsections (i.e. =head2, =head3, etc.)


=head1 SUBROUTINES/METHODS

A separate section listing the public components of the module's interface.
These normally consist of either subroutines that may be exported, or methods
that may be called on objects belonging to the classes that the module provides.
Name the section accordingly.

In an object-oriented module, this section should begin with a sentence of the
form "An object of this class represents...", to give the reader a high-level
context to help them understand the methods that are subsequently described.


=head1 DIAGNOSTICS

A list of every error and warning message that the module can generate
(even the ones that will "never happen"), with a full explanation of each
problem, one or more likely causes, and any suggested remedies.


=head1 CONFIGURATION AND ENVIRONMENT

A full explanation of any configuration system(s) used by the module,
including the names and locations of any configuration files, and the
meaning of any environment variables or properties that can be set. These
descriptions must also include details of any configuration language used.


=head1 DEPENDENCIES

A list of all the other modules that this module relies upon, including any
restrictions on versions, and an indication whether these required modules are
part of the standard Perl distribution, part of the module's distribution,
or must be installed separately.


=head1 INCOMPATIBILITIES

A list of any modules that this module cannot be used in conjunction with.
This may be due to name conflicts in the interface, or competition for
system or program resources, or due to internal limitations of Perl
(for example, many modules that use source code filters are mutually
incompatible).

=head1 SEE ALSO

L<...>

=head1 BUGS AND LIMITATIONS

A list of known problems with the module, together with some indication
whether they are likely to be fixed in an upcoming release.

Also a list of restrictions on the features the module does provide:
data types that cannot be handled, performance issues and the circumstances
in which they may arise, practical limitations on the size of data sets,
special cases that are not (yet) handled, etc.

The initial template usually just has:

There are no known bugs in this module.

Please report problems to Chris Weyl <cweyl@alumni.drew.edu>, or (preferred) 
to this package's RT tracker at E<bug-PACKAGE@rt.cpan.org>.

Patches are welcome.

=head1 AUTHOR

Chris Weyl  <cweyl@alumni.drew.edu>


=head1 LICENSE AND COPYRIGHT

Copyright (c) 2008 Chris Weyl <cweyl@alumni.drew.edu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the 

    Free Software Foundation, Inc.
    59 Temple Place, Suite 330
    Boston, MA  02111-1307  USA

=cut





t/lib/SingleCascade.pm
package SingleCascade;

use Moose;

has master => (
    #traits => [qw{Moose::Meta::Attribute::Custom::Trait::CascadeClear}], 
    traits => [qw{CascadeClearMaster}], 
    
    is => 'rw',
    isa => 'Str',
    clearer => 'clear_master',
    predicate => 'has_master',
    lazy => 1,
    default => 'nuts',
);

my @opts = (
    traits => [qw{CascadeClear}], 
    is => 'ro', isa => 'Str', clear_on => 'master', lazy_build => 1,
);

has sub1 => @opts; 
has sub2 => @opts;
has sub3 => @opts;

has nosub => (is => 'rw', isa => 'Str', lazy_build => 1);

sub _build_sub1 { shift->master . "1" }
sub _build_sub2 { shift->master . "2" }
sub _build_sub3 { shift->master . "3" }

sub _build_nosub { 'lazy!' }

1;



lib/MooseX/Traits/Attribute/CascadeClear.pm
#############################################################################
#
# Attribute metaclass trait to cascade clearers.
#
# Author:  Chris Weyl (cpan:RSRCHBOY), <cweyl@alumni.drew.edu>
# Company: No company, personal work
# Created: 12/24/2008 11:08:00 PM PST
#
# Copyright (c) 2008 Chris Weyl (cweyl@alumni.drew.edu) 
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
#############################################################################

package MooseX::Traits::Attribute::CascadeClear;

use Moose::Role;

our $VERSION = 0.03;

has clear_on => (
    is        => 'rw',
    isa       => 'Str',
    predicate => 'has_clear_on',
    clearer   => 'clear_clear_on',
);

no Moose::Role;

1;

__END__

=head1 NAME

MooseX::Traits::Attribute::CascadeClear - Attribute trait to cascade clearer actions

=head1 SYNOPSIS

    # our "master" attribute
    has master => (
        # overrides our clearer method 
        traits     => [ 'CascadeClearMaster' ], 
        is         => 'rw',
        isa        => 'Str',
        lazy       => 1,
        default    => 'nuts',
    );

    has client1 => (
        # mark us for clearing on clear_master()
        traits   => [ 'CascadeClear' ], 
        clear_on => 'master', 
        is       => 'ro', 
        isa      => 'Str', 
        lazy     => 1,
        default  => sub { shift->master . "1" }
    );

=head1 DESCRIPTION

CascadeClearer is an attribute metaclass trait that allows one to chain
attribute clearing actions.  For instance, if you have attributes that derive
their value from a "master" attribute, this would transparently allow you to
clear all the children attributes by clearing the master attribute.

=head1 USAGE

This trait provides an additional option to attributes, I<clear_on>.  It
should be used in conjunction with I<CascadeClearMaster> (aka
L<Moose::Meta::Attribute::Custom::Traits::CascadeClearMaster>).  As attributes
are defined, this trait should be specified and the clear_on attribute
declared.  Declaring a non-existant attribute in I<clear_on> causes neither an
error nor a warning.

To use, simply declare B<CascadeClear> as one of your attributes traits, then
designate its master attribute through I<clear_on>.  The client attribute does
not need to have a clearer defined, but if it does, it will be used.

=head1 SEE ALSO

L<MooseX::Traits::Attribute::CascadeClearMaster>

This package is part of the Fedora Project's Perl SIG work.  For more
information, see:

    http://fedoraproject.org
    http://fedoraproject.org/wiki/Perl
    http://camelus.fedorahosted.org

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems or requests to this package's RT tracker at 
<bug-MooseX-Traits-Attribute-CascadeClear@rt.cpan.org>.

Patches are welcome.

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>


=head1 LICENSE AND COPYRIGHT

Copyright (c) 2008 Chris Weyl <cweyl@alumni.drew.edu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=cut





lib/MooseX/Traits/Attribute/CascadeClearMaster.pm
#############################################################################
#
# Attribute metaclass trait to cascade clearers
#
# Author:  Chris Weyl (cpan:RSRCHBOY), <cweyl@alumni.drew.edu>
# Company: No company, personal work
# Created: 12/24/2008 11:08:00 PM PST
#
# Copyright (c) 2008 Chris Weyl <cweyl@alumni.drew.edu> 
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
#############################################################################

package MooseX::Traits::Attribute::CascadeClearMaster;

use Moose::Role;

our $VERSION = 0.03;

#use Smart::Comments '###', '####';

# I think this is kosher...
override process_accessors => sub {
    my ($self, $type, $accessor, $generate_as_inline_methods) = @_;

    # skip anything not a clearer
    return super if $type ne 'clearer';

    my $class = $self->associated_class;
    
    # I'm pretty sure the class/method_metaclass bit is correct...
    my $method = $class->method_metaclass->wrap(
        is_inline    => 0,     # right?
        package_name => $class->name,
        name         => $accessor, # $self->clearer,
        body         => sub { $self->_really_clear_value(shift) },
    );
    $self->associate_method($method);    

    return ($self->clearer, $method);
};
    
sub _really_clear_value {
    my ($self, $instance) = @_;

    # lifted from 
    #Class::MOP::Class
    #    ->initialize(ref($instance))
    #    ->get_meta_instance
    #    ->deinitialize_slot($instance, $self->name)
    #    ; 

    # clear our value 
    $self->clear_value($instance);

    # now, clear any other attribute that lists us as a master
    my $name = $self->name;
    my @atts = $instance->meta->get_all_attributes;

    for my $attribute (@atts) {

        ### working on: $attribute->name
        if ($attribute->does('CascadeClear') 
            && $attribute->has_clear_on
            && $attribute->clear_on eq $name) {

                if ($attribute->has_clearer) {

                    # this way cascading "just works"
                    my $att_clearer = $attribute->clearer;
                    $instance->$att_clearer();
                }
                else {

                    $attribute->clear_value($instance);
                }
        }
    }

}

no Moose::Role;

1;

__END__

=head1 NAME

MooseX::Traits::Attribute::CascadeClearMaster - Master clearer metatrait

=head1 SYNOPSIS

    # our "master" attribute
    has master => (

        # overrides our clearer method to check for dependent atts to clear
        traits     => [ 'CascadeClearMaster' ], 
        is         => 'rw',
        isa        => 'Str',
        lazy       => 1,
        default    => 'nuts',
    );

    has client1 => (

        # mark us for clearing on clear_master()
        traits   => [ 'CascadeClear' ], 
        clear_on => 'master', 

        is      => 'ro', 
        isa     => 'Str', 
        lazy    => 1,
        default => sub { shift->master . "1" }
    );

=head1 DESCRIPTION

CascadeClearer is an attribute metaclass trait that allows one to chain
attribute clearing actions.  For instance, if you have attributes that derive
their value from a "master" attribute, this would transparently allow you to
clear all the children attributes by clearing the master attribute.

=head1 USAGE

See L<MooseX::Traits::Attribute::CascadeClear>.

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems or requests to this package's RT tracker at 
<bug-MooseX-Traits-Attribute-CascadeClear@rt.cpan.org>.

Patches are welcome.

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>


=head1 LICENSE AND COPYRIGHT

Copyright (c) 2008 Chris Weyl <cweyl@alumni.drew.edu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=cut





